src/iosrv/sched.c

```diff
@@ -26,73 +26,953 @@
+{
+	D_ASSERT(len == sizeof(uuid_t));
+	return d_hash_string_u32((const char *)key, len);
+}
+
+static void
+spi_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct sched_pool_info	*spi = sched_rlink2spi(rlink);
+
+	spi->spi_ref++;
+}
+
+static bool
+spi_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct sched_pool_info	*spi = sched_rlink2spi(rlink);
+
+	D_ASSERT(spi->spi_ref > 0);
+	spi->spi_ref--;
+
+	return spi->spi_ref == 0;
+}
+
+static void
+spi_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct sched_pool_info	*spi = sched_rlink2spi(rlink);
+	unsigned int		 type;
+
+	for (type = SCHED_REQ_IO; type < SCHED_REQ_MAX; type++) {
+		D_ASSERT(pool2req_cnt(spi, type) == 0);
+		D_ASSERT(d_list_empty(pool2req_list(spi, type)));
+	}
+
+	D_FREE(spi);
+}
+
+static d_hash_table_ops_t sched_pool_hash_ops = {
+	.hop_key_cmp	= spi_key_cmp,
+	.hop_key_hash	= spi_key_hash,
+	.hop_rec_addref	= spi_rec_addref,
+	.hop_rec_decref	= spi_rec_decref,
+	.hop_rec_free	= spi_rec_free,
+};
+
+/*
+ * d_hash_table_traverse() doesn't support item deletion in traverse
+ * callback, so the stale 'spi' (pool was destroyed) will be added into
+ * a purge list in traverse callback and being deleted later.
+ */
+struct purge_item {
+	d_list_t	pi_link;
+	uuid_t		pi_pool_id;
+};
+
+static void
+prune_purge_list(struct dss_xstream *dx)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct purge_item	*pi, *tmp;
+	bool			 deleted;
+
+	d_list_for_each_entry_safe(pi, tmp, &info->si_purge_list, pi_link) {
+		deleted = d_hash_rec_delete(info->si_pool_hash, pi->pi_pool_id,
+					    sizeof(uuid_t));
+		if (!deleted)
+			D_ERROR("XS(%d): Purge "DF_UUID" failed.\n",
+				dx->dx_xs_id, DP_UUID(pi->pi_pool_id));
+
+		d_list_del_init(&pi->pi_link);
+		D_FREE(pi);
+	}
+}
+
+static void
+add_purge_list(struct dss_xstream *dx, struct sched_pool_info *spi)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct purge_item	*pi;
+
+	D_ALLOC_PTR(pi);
+	if (pi == NULL) {
+		D_ERROR("XS(%d): Alloc purge item failed.\n", dx->dx_xs_id);
+		return;
+	}
+	D_INIT_LIST_HEAD(&pi->pi_link);
+	uuid_copy(pi->pi_pool_id, spi->spi_pool_id);
+	d_list_add_tail(&pi->pi_link, &info->si_purge_list);
+}
+
+static void
+sched_info_fini(struct dss_xstream *dx)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_request	*req, *tmp;
+
+	D_ASSERT(info->si_req_cnt == 0);
+	D_ASSERT(d_list_empty(&info->si_sleep_list));
+	D_ASSERT(d_list_empty(&info->si_fifo_list));
+
+	prune_purge_list(dx);
+
+	if (info->si_pool_hash) {
+		d_hash_table_destroy(info->si_pool_hash, true);
+		info->si_pool_hash = NULL;
+	}
+
+	d_list_for_each_entry_safe(req, tmp, &info->si_idle_list,
+				   sr_link) {
+		d_list_del_init(&req->sr_link);
+		D_FREE(req);
+	}
+}
+
+static int
+prealloc_requests(struct sched_info *info, int cnt)
+{
+	struct sched_request	*req;
+	int			 i;
+
+	for (i = 0; i < cnt; i++) {
+		D_ALLOC_PTR(req);
```

Comment:
My intention is to avoid in-flight allocation and make all requests allocated during init time, so allocation performance isn't critical here. D_ALLOC_ARRAY() could be faster, but it will end up with allocating a huge array. Let's keep current manner for this moment to see if can work well.